### PR TITLE
Implement Configuration stage of Builder

### PR DESCRIPTION
**Pull request by Themanwhosmellslikesugar**

### CODE REVIEW

File: src/common/components/Input.jsx [hunk 1]
```diff
@@ -1,51 +1,59 @@
...
-                type={type}
-                className={`form-control ${isValidClass}`}
-                placeholder={placeholder}
-                name={name}
-                value={value}
+            <DebounceInput
+                type={this.props.type}
+                className={"form-control " + (!isValid && invalidCssClass)}
+                id={this.props.id}
+                debounceTimeout={600}
```
Comment (karpa4o4) [inline src/common/components/Input.jsx]: Let's define `debounceTimeout` value as constat.

File: src/modules/Builder/components/Builder/Builder.jsx [hunk 2]
```diff
@@ -39,20 +40,28 @@ export default class Builder extends Component {
     }
 
     onNextBuildState() {
+        this.builderCallback = this.builderCallback.bind(this);
         if (!this.waitExecutingState) {
             this.waitExecutingState = true;
-
             this.currentStateRef.current.executeState((stateData) => {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Builder/Builder.jsx]: Anomymous function is redunant.

File: src/modules/Builder/components/Builder/Builder.jsx [hunk 3]
```diff
@@ -39,20 +40,28 @@ export default class Builder extends Component {
...
+    builderCallback(stateData) {
+        this.processStateData(stateData);
+    }
+
     processStateData(data) {
         const { currentRoute, prevRouteIndex } = this.state;
         switch (data ? data.state : "") {
             case "initialization":
                 this.buildUUID = data.buildUUID;
+                this.setState(({ device: data.device, os: data.os }));
```
Comment (karpa4o4) [inline src/modules/Builder/components/Builder/Builder.jsx]: Let's set the state asynchronously everywhere. Look here: https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous

File: src/modules/Builder/components/Builder/Builder.jsx [hunk 4]
```diff
@@ -124,12 +141,13 @@ export default class Builder extends Component {
                                     <div className="row">
                                         <div className="col-md-9">{" "}</div>
                                         <div className="col-md-3">
-                                            <Button
-                                                styleName="btn-primary btn-block next-state"
+                                            <button
```
Comment (karpa4o4) [inline src/modules/Builder/components/Builder/Builder.jsx]: Removing the `Button` component should be made a separate commit.

File: src/modules/Builder/components/Builder/Builder.jsx [hunk 5]
```diff
@@ -8,7 +8,6 @@ import {
 } from "react-router-dom";
 
 import SidebarPage from "common/containers/SidebarPage";
-import Button from "common/components/Button";
```
Comment (karpa4o4) [inline src/modules/Builder/components/Builder/Builder.jsx]: Removing the `Button` component should be made a separate commit.

File: src/modules/Builder/components/Builder/config.js [hunk 6]
```diff
@@ -8,20 +9,20 @@ const routes = [
...
-        path: `${routeBasePath}/packagelist`,
-        title: "Package list",
-        main: PackageList,
-    },
-    {
-        path: `${routeBasePath}/usergroups`,
-        title: "User and groups",
-        main: "",
-    },
+    // {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Builder/config.js]: This redunant changes.

File: src/modules/Builder/components/Builder/config.js [hunk 7]
```diff
@@ -8,20 +9,20 @@ const routes = [
         title: "Initialization",
         main: Initialization,
     },
-    {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Builder/config.js]: This redunant changes.

File: src/modules/Builder/components/Configuration/Configuration.jsx [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+export default class Configuration extends Component {
+    constructor(props) {
+        super(props);
+        
+        this.buttonStateCallback = this.buttonStateCallback.bind(this);
+    }
+
+    executeState(builderCallback) {
+        this.builderCallback = builderCallback.bind(this);
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/Configuration.jsx]: Define `builderCallback` in constructor. For example: https://github.com/tolstoyevsky/cusdeb-web-client/blob/master/src/modules/Builder/components/Initialization/Initialization.jsx#L25

File: src/modules/Builder/components/Configuration/components/Network/style.scss [hunk 9]
```diff
@@ -0,0 +1,3 @@
+.custom-control {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Network/style.scss]: You need to make that these styles are applied only in the `Network` component. Otherwise, they will be applied in other components.

File: src/common/components/Input.jsx [hunk 10]
```diff
@@ -35,12 +45,15 @@ export default class Input extends Component {
...
 Input.propTypes = {
+    id: PropTypes.string,
     isValid: PropTypes.bool,
     name: PropTypes.string,
     onChange: PropTypes.func,
     placeholder: PropTypes.string,
     type: PropTypes.string.isRequired,
+    validationFunc: PropTypes.func,
     value: PropTypes.string,
+    disabled: PropTypes.bool,
```
Comment (karpa4o4) [inline src/common/components/Input.jsx]: This change must be in the next commit: https://github.com/tolstoyevsky/cusdeb-web-client/pull/16/commits/a48bf13e3b229b89654a74b7a7ba3b6d09d39b45

File: src/modules/Builder/components/PackageList/PackageList.jsx [hunk 11]
```diff
@@ -46,6 +47,11 @@ export default class PackageList extends Component {
             });
     }
 
+    executeState(builderCallback) {
+        this.builderCallback = builderCallback.bind(this);
```
Comment (karpa4o4) [inline src/modules/Builder/components/PackageList/PackageList.jsx]: It's enought to call the function passed to executeState method, without defining is as an attribute of the class instance.

File: src/modules/Builder/components/Builder/Builder.jsx [hunk 12]
```diff
@@ -38,20 +40,31 @@ export default class Builder extends Component {
     }
 
     onNextBuildState() {
+        this.builderCallback = this.builderCallback.bind(this);
```
Comment (karpa4o4) [inline src/modules/Builder/components/Builder/Builder.jsx]: You shoud bind this function in constructor.

File: src/modules/Builder/components/Configuration/components/Network/Network.jsx [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+            value: item,
+            text: item,
+        }));
+    }
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            hostName: "",
+            timeZone: "Europe/London",
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Network/Network.jsx]: Please, specify default value like as `timeZones[0]`.

File: src/modules/Builder/components/Configuration/components/WirelessSupport/WirelessSupport.jsx [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+                "SSID": SSID,
+                "PSK": PSK,
+            }
+        );
+    }
+
+    componentDidMount() {
+        RPC.fetchDefaultConfig()
+            .then((defaultConfig) => {
+                this.setState(() => ({ defaultConfig, SSID: defaultConfig.SSID, PSK: defaultConfig.PSK }));
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/WirelessSupport/WirelessSupport.jsx]: Why do you save `defaultConfig`?

File: src/modules/Builder/components/Initialization/Initialization.jsx [hunk 15]
```diff
@@ -75,6 +76,8 @@ export default class Initialization extends Component {
                 this.builderCallback({
                     state: "initialization",
                     buildUUID: event,
+                    device: selectsData.device,
```
Comment (karpa4o4) [inline src/modules/Builder/components/Initialization/Initialization.jsx]: Let's put the changes in this file in a separate commit.

File: src/styles/index.scss [hunk 16]
```diff
@@ -8,3 +8,7 @@
 @import "~modules/Builder/components/PackageList/components/TotalPackagesPanel/styles.scss";
 
 @import "~modules/Error404/components/Error404/styles.scss";
+
+@import "~modules/Builder/components/Configuration/style.scss";
```
Comment (karpa4o4) [inline src/styles/index.scss]: Move these lines to the Builder style imports.

**Comment by karpa4o4** [comment]
@Themanwhosmellslikesugar, let's document the required steps to consider when adding a new configuration section.

File: src/api/rpc/blackmagic.js [hunk 17]
```diff
@@ -40,3 +42,11 @@ export const searchPackages = async (packageName) => (
 export const resolvePackages = async (packageList) => (
     rpcClient.emit(RESOLVE_PACKAGES_PR, packageList)
 );
+
+export const fetchDefaultConfig = async () => (
```
Comment (karpa4o4) [inline src/api/rpc/blackmagic.js]: Let's rename it to `fetchDefaultConfigurationParams`.

File: src/common/components/Input.jsx [hunk 18]
```diff
@@ -1,6 +1,9 @@
 import React, { Component } from "react";
 import PropTypes from "prop-types";
+import { DebounceInput } from "react-debounce-input";
 
+const invalidCssClass = "is-invalid";
+const DEBOUNCE_TIMEOUT = 600;
```
Comment (karpa4o4) [inline src/common/components/Input.jsx]: Add an empty string after this string.

File: src/modules/Builder/components/Configuration/components/Network/Network.jsx [hunk 19]
```diff
@@ -0,0 +1,124 @@
+import React, { Component } from "react";
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Network/Network.jsx]: Let's rename this component to `Main`.

File: src/modules/Builder/components/Builder/Builder.jsx [hunk 20]
```diff
@@ -42,17 +44,26 @@ export default class Builder extends Component {
...
         const { currentRoute, prevRouteIndex } = this.state;
         switch (data ? data.state : "") {
             case "initialization":
                 this.buildUUID = data.buildUUID;
+                this.setState(() => ({
+                    device: data.device,
+                    os: data.os,
+                }));
+                break;
+            case "button state":
```
Comment (karpa4o4) [inline src/modules/Builder/components/Builder/Builder.jsx]: Let's rename it to `buttonState`.

File: src/modules/Builder/components/Configuration/Configuration.jsx [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+export default class Configuration extends Component {
+    constructor(props) {
+        super(props);
+
+        this.serverRefs = {
+            network: React.createRef(),
+            wirelessSupport: React.createRef(),
+        };
+
+        this.buttonStates = {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/Configuration.jsx]: `buttonStates` needs to be an attribute of the component state.

File: src/modules/Builder/components/Configuration/components/WirelessSupport/WirelessSupport.jsx [hunk 22]
```diff
@@ -0,0 +1,165 @@
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+
+import Card from "common/containers/Card";
+import Input from "common/components/Input";
+import InputGroup from "common/components/InputGroup";
+
+import * as RPC from "api/rpc/blackmagic";
+
+export default class WirelessSupport extends Component {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/WirelessSupport/WirelessSupport.jsx]: It's better to call this component simply `Wireless`.

File: src/modules/Builder/components/Configuration/config.js [hunk 23]
```diff
@@ -0,0 +1,10 @@
+const suitableOS = [
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/config.js]: The name `wirelessSupportOS` is more approprivate for this variable.

File: src/modules/Builder/components/Configuration/config.js [hunk 24]
```diff
@@ -0,0 +1,10 @@
...
+    "Raspbian 10 \"Buster\" (32-bit)",
+];
+
+const suitableDevices = [
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/config.js]: The name `wirelessSupportDevices` is more approprivate for this variable.

File: src/modules/Builder/components/Configuration/Configuration.jsx [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+            builderCallback({
+                state: "button state",
+                buttonState: true,
+            });
+        }
+    }
+
+    checkWirelessSupport() {
+        const { os, device } = this.props;
+        if (suitableOS.includes(os) && suitableDevices.includes(device)) return true;
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/Configuration.jsx]: It's better to use brackets everywhere. This way the code loks more expressive.

File: src/modules/Builder/components/Configuration/Configuration.jsx [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+            wirelessSupportPSK: true,
+        };
+
+        this.builderCallback = () => { };
+
+        this.buttonStateCallback = this.buttonStateCallback.bind(this);
+    }
+
+    executeState(builderCallback) {
+        this.builderCallback = builderCallback.bind(this);
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/Configuration.jsx]: You don't need create property in `this`, you cat take `builderCallback` from props.

File: src/modules/Builder/components/PackageList/PackageList.jsx [hunk 27]
```diff
@@ -46,6 +48,11 @@ export default class PackageList extends Component {
             });
     }
 
+    executeState(builderCallback) {
+        this.builderCallback = this.builderCallback.bind(this);
```
Comment (karpa4o4) [inline src/modules/Builder/components/PackageList/PackageList.jsx]: It's redunant changes.

File: src/modules/Builder/components/PackageList/PackageList.jsx [hunk 28]
```diff
@@ -16,6 +16,8 @@ export default class PackageList extends Component {
             dependentPackages: [],
         };
 
+        this.builderCallback = () => { };
```
Comment (karpa4o4) [inline src/modules/Builder/components/PackageList/PackageList.jsx]: It's redunant changes.

File: src/modules/Builder/components/Configuration/components/Network/Network.jsx [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    componentDidMount() {
+        RPC.fetchDefaultConfig()
+            .then((defaultConfig) => {
+                this.setState(() => ({ hostName: defaultConfig.hostname }));
+            });
+    }
+
+    onInputChange(name, value) {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Network/Network.jsx]: Let's rename it to `onHostNameChange`.

File: src/modules/Builder/components/Configuration/components/Network/Network.jsx [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+            text: item,
+        }));
+    }
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            hostName: "",
+            timeZone: timeZones[0],
+            warningStatus: false,
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Network/Network.jsx]: How `warningStatus` are used in this component?

File: src/modules/Builder/components/Configuration/components/Network/Network.jsx [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    onTimeZoneChange(value) {
+        const timeZone = timeZones.find((item) => item === value);
+        this.setState(() => ({
+            timeZone,
+        }));
+    }
+
+    hostNameValidor(value) {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Network/Network.jsx]: You need to make sure that all validation functions have one setState call and one return.

File: src/modules/Builder/components/Configuration/styles.scss [hunk 32]
```diff
@@ -0,0 +1,9 @@
+.builder-configuration-stage {
+  .wifi-group {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/styles.scss:2]: This styles is redunant, because `width: 100%;` exist in `input-group` class.

File: src/modules/Builder/components/Configuration/components/Wireless/styles.scss [hunk 33]
```diff
@@ -0,0 +1,13 @@
+.configuration-wireless {
+  .form-control-feedback {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Wireless/styles.scss:2]: Where is class is used?

File: src/modules/Builder/components/Configuration/config.js [hunk 34]
```diff
@@ -0,0 +1,10 @@
+const wirelessSupportOS = [
+    "Raspbian 10 \"Buster\" (32-bit)",
```
[the remark this answers is not in the document]
Comment (Themanwhosmellslikesugar) [reply src/modules/Builder/components/Configuration/config.js:2]: eslint requires that strings must use doublequote

[the remark this answers is not in the document]
Comment (karpa4o4) [reply src/modules/Builder/components/Configuration/config.js:2]: To solve this problem you need to add [allowtemplateliterals](https://eslint.org/docs/rules/quotes#allowtemplateliterals) to `.eslintrc.js`.

File: src/modules/Builder/components/Builder/config.js [hunk 35]
```diff
@@ -1,4 +1,5 @@
 import Initialization from "modules/Builder/components/Initialization/Initialization";
+import Configuration from "modules/Builder/components/Configuration/Configuration";
```
Comment (karpa4o4) [inline src/modules/Builder/components/Builder/config.js:2]: Please, keep the alphabetical order of impors.

File: src/modules/Builder/components/Configuration/Configuration.jsx [hunk 36]
```diff
@@ -0,0 +1,85 @@
...
+        builderCallback();
+        Object.keys(this.serverRefs).forEach((key) => {
+            this.serverRefs[key].current.syncConfigParams();
+        });
+    }
+
+    buttonStateCallback(name, buttonVariable) {
+        const { builderCallback } = this.props;
+        const { buttonStates } = this.state;
+        buttonStates[name] = buttonVariable;
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/Configuration.jsx:37]: Please, use `setState`, because this expression is an antipattern.

File: src/modules/Builder/components/Configuration/Configuration.jsx [hunk 37]
```diff
@@ -0,0 +1,85 @@
...
+    render() {
+        return (
+            <div className="builder-configuration-stage" key="cards">
+                <Main
+                    buttonStateCallback={this.buttonStateCallback}
+                    ref={this.serverRefs.main}
+                />
+                {(this.checkWirelessSupport())
+                    && (
+                        (
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/Configuration.jsx:69]: This pair of brackets is unnecessary.

File: src/modules/Builder/components/Configuration/Configuration.jsx [hunk 38]
```diff
@@ -0,0 +1,85 @@
...
+                                ref={this.serverRefs.wireless}
+                            />
+                        )
+                    )}
+            </div>
+        );
+    }
+}
+
+Configuration.propTypes = {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/Configuration.jsx:81]: Add these props to all components of the builder stages.

File: src/modules/Builder/components/Configuration/components/Main/Main.jsx [hunk 39]
```diff
@@ -0,0 +1,124 @@
...
+            text: item,
+        }));
+    }
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            hostName: "",
+            timeZone: timeZones[0],
+            previosHostName: "",
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Main/Main.jsx:26]: How is this state attribute used?

File: src/modules/Builder/components/Configuration/components/Wireless/Wireless.jsx [hunk 40]
```diff
@@ -0,0 +1,154 @@
...
+import PropTypes from "prop-types";
+
+import Card from "common/containers/Card";
+import Input from "common/components/Input";
+import InputGroup from "common/components/InputGroup";
+
+import * as RPC from "api/rpc/blackmagic";
+
+export default class Wireless extends Component {
+    static WarningMessege() {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Wireless/Wireless.jsx:11]: Name of methods should start with a small latter.

File: src/modules/Builder/components/Configuration/components/Wireless/Wireless.jsx [hunk 41]
```diff
@@ -0,0 +1,154 @@
...
+                Пароль должен быть пустой либо содержать от 8 до 63 символов,
+                иначе сборка завершится неудачно.
+            </div>
+        );
+    }
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            PSK: "",
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Wireless/Wireless.jsx:23]: Separate the form attributes from other state attributes with a paragraph.

File: src/modules/Builder/components/Configuration/components/Wireless/Wireless.jsx [hunk 42]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    componentDidMount() {
+        RPC.fetchDefaultConfigurationParams()
+            .then((defaultConfig) => {
+                this.setState(() => ({ SSID: defaultConfig.SSID }));
+            });
+    }
+
+    onCheckboxChange(value) {
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Wireless/Wireless.jsx:43]: * You can use one `setState`;
* `disabledState` duplicates the assignment of `enableWireless`, you can get rid of it.

File: src/modules/Builder/components/Configuration/components/Wireless/Wireless.jsx [hunk 43]
```diff
@@ -0,0 +1,154 @@
...
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            PSK: "",
+            SSID: "",
+            disabledState: true,
+            warningStatus: false,
+            enableWireless: true,
+            previosPSK: "",
```
Comment (karpa4o4) [inline src/modules/Builder/components/Configuration/components/Wireless/Wireless.jsx:28]: How is this state attribute used?

File: src/modules/Builder/components/Configuration/Configuration.jsx (hunk 38 above)
Comment (Themanwhosmellslikesugar) [reply src/modules/Builder/components/Configuration/Configuration.jsx:81]: PackageList stage doesnt use these props, hence no verification is needed, eslint points on this
